src/appleseed/renderer/kernel/lighting/lighttree.cpp

```diff
@@ -0,0 +1,333 @@
+                                               : "color.green";
+
+            const auto& bbox_left = m_nodes[i].get_left_bbox();
+            const auto& bbox_right = m_nodes[i].get_right_bbox();
+
+            file.draw_aabb(bbox_left, color, Width);
+            file.draw_aabb(bbox_right, color, Width);
+        }
+    }
+}
+
+//
+// Calculate the tree depth and assign total luminance to each node of the tree.
+// Total luminance represents the sum of all its child nodes luminances.
+//
+
+float LightTree::recursive_node_update(size_t node_index, size_t node_level)
+{
+    float luminance = 0.0f;
+
+    if (!m_nodes[node_index].is_leaf())
+    {
+        const auto& child1 = m_nodes[node_index].get_child_node_index();
+        const auto& child2 = m_nodes[node_index].get_child_node_index() + 1;
+
+        float luminance1 = recursive_node_update(child1, node_level + 1);
+        float luminance2 = recursive_node_update(child2, node_level + 1);
+
+        luminance = luminance1 + luminance2;
+    }
+    else
+    {
+        // Access the light intensity value.
+        size_t item_index = m_nodes[node_index].get_item_index();
+        size_t light_source_index = m_items[item_index].m_light_sources_index;
+        Spectrum spectrum = m_light_sources[light_source_index]->get_intensity();
+
+        // Luminance is the average accross all the spectrum channels
+        for(size_t i = 0; i < spectrum.size(); i++)
+        {
+            luminance += spectrum[i];
+        }
+        luminance /= spectrum.size();
+
+        // Modify the tree depth if the branch is deeper than the other branches
+        // visited so far
+        if (m_tree_depth < node_level)
+            m_tree_depth = node_level;
+    }
+
+    m_nodes[node_index].set_luminance(luminance);
+    m_nodes[node_index].set_level(node_level);
+
+    return luminance;
+}
+
+std::pair<size_t, float> LightTree::sample(
+        const foundation::Vector3d    surface_point,
+        float                         s) const
+{
+    float light_probability = 1.0;
+    size_t node_index = 0;
+
+    while (!m_nodes[node_index].is_leaf())
+    {
+        // LightTreeNode
+        const auto& node = m_nodes[node_index];
+
+        std::pair<float, float> result = child_node_probabilites(node, surface_point);
+        const float p1 = result.first;
+        const float p2 = result.second;
+
+        if (s <= p1)
+        {
+            light_probability *= p1;
+            s /= p1;
+            node_index = node.get_child_node_index();
+        }
+        else
+        {
+            light_probability *= p2;
+            s = (s - p1) / p2;
+            node_index = node.get_child_node_index() + 1;
+        }
+    }
+    size_t item_index = m_nodes[node_index].get_item_index();
+    size_t light_index = m_items[item_index].m_light_tree_external_index;
+
+    return std::pair<size_t, float>(light_index, light_probability);
+}
+
+std::pair<float, float> LightTree::child_node_probabilites(
```

Comment:
The code would be simpler and especially clearer if this function was inlined where it is called (i.e. if the function wouldn't exist as a separate entity).